### DISCUSSION TITLE
How can I restyle a NumberBox control?

### DISCUSSION THREAD

**Discussion by AndrewKeepCoding** [microsoft/microsoft-ui-xaml#7764] [Q&A] [answered]
I have a custom control, ``CustomNumberBox`` derived from ``NumberBox``, and I want to restyle its control template but the style doesn't get applied. AFAICS, this is the same on all controls under the ``Microsoft.UI.Xaml.Controls`` namespace. 

- ``Breadcrumbar``
- ``ColorPicker``
- ``InfoBar``
- ``NumberBox``
- ``ProgressBar``/``ProgressRing``
- ``RatingControl``

So, apparently this is by design but is there any way to restyle these controls?

**Comment by marcelwgn** [comment, marked as answer]
How did you restyle the control and where did you place the template?

**Reply by AndrewKeepCoding** [reply]
The style is a simple style just to check if it gets applied.

**Themes/Generic.xaml**
```xaml
<Style TargetType="local:CustomNumberBox">
    <Setter Property="Template">
        <Setter.Value>
            <ControlTemplate TargetType="local:CustomNumberBox">
                <Border
                    Background="{TemplateBinding Background}"
                    BorderBrush="{TemplateBinding BorderBrush}"
                    BorderThickness="{TemplateBinding BorderThickness}">
                    <TextBlock Text="CustomNumberBoxStyle" />
                </Border>
            </ControlTemplate>
        </Setter.Value>
    </Setter>
</Style>
```

**CustomNumberBox.cs**
```csharp
public sealed class CustomNumberBox : NumberBox  /*Control works*/
{
    public CustomNumberBox()
    {
        this.DefaultStyleKey = typeof(CustomNumberBox);
    }
}
```

The style does get applied (renders the ``TextBlock`` in the style) if ``CustomNumberBox`` derives from ``Control`` instead.

**Reply by marcelwgn** [reply]
I think the issue is that the way the controls apply their template in WinUI 2 is a bit more complicated. They use this lookup logic to find the right template: https://github.com/microsoft/microsoft-ui-xaml/blob/da8b658b4a401ed11103b115fc70a82427184fb8/dev/dll/XamlControlsResources.cpp#L266-L364

I was able to get this working by putting the template in the App.xaml file as part of the merged dictionaries.

**Reply by AndrewKeepCoding** [reply]
Placing the style in the App.xaml worked! Thanks @chingucoding.

**Reply by AndrewKeepCoding** [reply]
This is just for clarification but, this is a workaround right? You can't use this way on control libraries separated from the app. Should I create an issue?

**Reply by marcelwgn** [reply]
Yes @AndrewKeepCoding , this is more or less a workaround. A normal library could probably work around, I'm inclined to say this issue might be caused by the "magic" around the `Themes/Generic.xaml`. Loading a resource dictionary by referencing the XAML file would probably work just as expected.

**Reply by 4broadcast** [reply]
Hi, does anyone know if there is an open issue somewhere for this? I just found this question, which seems to be the same as mine here https://github.com/microsoft/microsoft-ui-xaml/discussions/10196.
